File: backend/app/modules/bolt/patch_applier.py
"""
Unified Diff Patch Applier
Applies Git-style patches to files (Python backend version)
"""

import re
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass


@dataclass
class DiffHunk:
    old_start: int
    old_lines: int
    new_start: int
    new_lines: int
    changes: List[Dict[str, str]]


@dataclass
class ParsedDiff:
    old_file: str
    new_file: str
    hunks: List[DiffHunk]


def parse_unified_diff(patch: str) -> ParsedDiff:
    """Parse a unified diff patch"""
    lines = patch.strip().split('\n')

    diff = ParsedDiff(old_file="", new_file="", hunks=[])

    current_hunk = None
    i = 0

    while i < len(lines):
        line = lines[i]

        # Parse file headers
        if line.startswith('--- '):
            diff.old_file = line[4:].strip().replace('a/', '')
        elif line.startswith('+++ '):
            diff.new_file = line[4:].strip().replace('b/', '')

        # Parse hunk header
        elif line.startswith('@@'):
            # @@ -10,3 +10,7 @@
            match = re.search(r'@@ -(\d+),?(\d*) \+(\d+),?(\d*) @@', line)
            if match:
                old_start = int(match.group(1))
                old_lines = int(match.group(2)) if match.group(2) else 1
                new_start = int(match.group(3))
                new_lines = int(match.group(4)) if match.group(4) else 1

                if current_hunk:
                    diff.hunks.append(current_hunk)

                current_hunk = DiffHunk(
                    old_start=old_start,
                    old_lines=old_lines,
                    new_start=new_start,
                    new_lines=new_lines,
                    changes=[]
                )

        # Parse changes
        elif current_hunk is not None:
            if line.startswith('+') and not line.startswith('+++'):
                current_hunk.changes.append({
                    'type': 'add',
                    'content': line[1:]
                })
            elif line.startswith('-') and not line.startswith('---'):
                current_hunk.changes.append({
                    'type': 'remove',
                    'content': line[1:]
                })
            elif line.startswith(' '):
                current_hunk.changes.append({
                    'type': 'context',
                    'content': line[1:]
                })

        i += 1

    # Add last hunk
    if current_hunk:
        diff.hunks.append(current_hunk)

    return diff


def apply_unified_patch(
    original_content: str,
    patch: str
) -> Dict:
    """
    Apply a unified diff patch to original content

    Args:
        original_content: Original file content
        patch: Unified diff patch

    Returns:
        Dict with success, new_content, error, conflicts
    """
    try:
        diff = parse_unified_diff(patch)
        lines = original_content.split('\n')

        # Apply each hunk
        for hunk in diff.hunks:
            result = _apply_hunk(lines, hunk)
            if not result['success']:
                return result

        return {
            'success': True,
            'new_content': '\n'.join(lines),
            'error': None,
            'conflicts': None
        }

    except Exception as e:
        return {
            'success': False,
            'new_content': None,
            'error': f"Failed to apply patch: {str(e)}",
            'conflicts': None
        }


def _apply_hunk(lines: List[str], hunk: DiffHunk) -> Dict:
    """Apply a single hunk to the lines array"""
    # Convert to 0-based index
    current_line = hunk.old_start - 1
    original_lines = lines.copy()

    try:
        for change in hunk.changes:
            change_type = change['type']
            content = change['content']

            if change_type == 'context':
                # Verify context line matches
                if current_line >= len(lines):
                    return {
                        'success': False,
                        'error': f"Context mismatch at line {current_line + 1}: exceeded file length",
                        'conflicts': [
                            f"Line {current_line + 1}",
                            "File is shorter than expected"
                        ]
                    }

                if lines[current_line] != content:
                    return {
                        'success': False,
                        'error': f"Context mismatch at line {current_line + 1}. Expected: \"{content}\", Got: \"{lines[current_line]}\"",
                        'conflicts': [
                            f"Line {current_line + 1}",
                            f"Expected: {content}",
                            f"Got: {lines[current_line]}"
                        ]
                    }
                current_line += 1

            elif change_type == 'remove':
                # Verify line to remove matches
                if current_line >= len(lines):
                    return {
                        'success': False,
                        'error': f"Cannot remove line {current_line + 1}: exceeded file length",
                        'conflicts': [
                            f"Line {current_line + 1}",
                            "File is shorter than expected"
                        ]
                    }

                if lines[current_line] != content:
                    return {
                        'success': False,
                        'error': f"Cannot remove line {current_line + 1}. Content mismatch.",
                        'conflicts': [
                            f"Line {current_line + 1}",
                            f"Expected to remove: {content}",
                            f"Got: {lines[current_line]}"
                        ]
                    }
                # Remove the line
                lines.pop(current_line)

            elif change_type == 'add':
                # Insert new line
                lines.insert(current_line, content)
                current_line += 1

        return {'success': True}

    except Exception as e:
        # Restore original lines on error
        lines.clear()
        lines.extend(original_lines)

        return {
            'success': False,
            'error': f"Hunk application failed: {str(e)}"
        }


def apply_patch_fuzzy(
    original_content: str,
    patch: str,
    fuzziness: int = 2
) -> Dict:
    """
    Apply patch with fuzzy matching

    Args:
        original_content: Original file content
        patch: Unified diff patch
        fuzziness: Number of lines to search around the specified location

    Returns:
        Dict with success, new_content, error, conflicts
    """
    # Try exact match first
    result = apply_unified_patch(original_content, patch)
    if result['success']:
        return result

    # Try fuzzy matching
    try:
        diff = parse_unified_diff(patch)
        lines = original_content.split('\n')

        for hunk in diff.hunks:
            # Try exact match
            result = _apply_hunk(lines.copy(), hunk)

            if not result['success']:
                # Try fuzzy matching with offset
                found = False
                for offset in range(-fuzziness, fuzziness + 1):
                    if offset == 0:
                        continue

                    modified_hunk = DiffHunk(
                        old_start=hunk.old_start + offset,
                        old_lines=hunk.old_lines,
                        new_start=hunk.new_start + offset,
                        new_lines=hunk.new_lines,
                        changes=hunk.changes
                    )

                    result = _apply_hunk(lines.copy(), modified_hunk)
                    if result['success']:
                        _apply_hunk(lines, modified_hunk)
                        found = True
                        break

                if not found:
                    return {
                        'success': False,
                        'error': f"Could not find matching context for hunk (tried fuzziness: {fuzziness})",
                        'new_content': None,
                        'conflicts': None
                    }

        return {
            'success': True,
            'new_content': '\n'.join(lines),
            'error': None,
            'conflicts': None
        }

    except Exception as e:
        return {
            'success': False,
            'new_content': None,
            'error': f"Fuzzy patch failed: {str(e)}",
            'conflicts': None
        }


def reverse_patch(patch: str) -> str:
    """Create a reverse patch for undo operations"""
    lines = patch.strip().split('\n')
    reversed_lines = []

    for line in lines:
        if line.startswith('+++'):
            # Swap +++ and ---
            reversed_lines.append(line.replace('+++', '---', 1))
        elif line.startswith('---'):
            reversed_lines.append(line.replace('---', '+++', 1))
        elif line.startswith('+') and not line.startswith('+++'):
            # + becomes -
            reversed_lines.append('-' + line[1:])
        elif line.startswith('-') and not line.startswith('---'):
            # - becomes +
            reversed_lines.append('+' + line[1:])
        else:
            # Context lines and headers remain the same
            reversed_lines.append(line)

    return '\n'.join(reversed_lines)


def preview_patch(original_content: str, patch: str) -> Dict:
    """
    Preview patch changes without applying

    Returns:
        Dict with additions, deletions, changes
    """
    try:
        diff = parse_unified_diff(patch)

        additions = 0
        deletions = 0
        changes = []

        for hunk in diff.hunks:
            line_num = hunk.new_start

            for change in hunk.changes:
                if change['type'] == 'add':
                    additions += 1
                    changes.append({
                        'line': line_num,
                        'type': 'add',
                        'content': change['content']
                    })
                    line_num += 1
                elif change['type'] == 'remove':
                    deletions += 1
                    changes.append({
                        'line': line_num,
                        'type': 'remove',
                        'content': change['content']
                    })
                else:
                    line_num += 1

        return {
            'additions': additions,
            'deletions': deletions,
            'changes': changes
        }

    except Exception as e:
        return {
            'additions': 0,
            'deletions': 0,
            'changes': [],
            'error': str(e)
        }

File: frontend/src/services/diffParser/patchApplier.ts
/**
 * Patch Applier - Applies unified diffs to files
 * Core component for AI code modifications
 */

import { ParsedDiff, DiffHunk, parseDiff } from './patchParser'

export interface PatchResult {
  success: boolean
  newContent?: string
  error?: string
  conflicts?: string[]
}

/**
 * Apply a unified diff patch to original content
 */
export function applyPatch(
  originalContent: string,
  patch: string | ParsedDiff
): PatchResult {
  try {
    const diff = typeof patch === 'string' ? parseDiff(patch) : patch
    const lines = originalContent.split('\n')

    // Apply each hunk
    for (const hunk of diff.hunks) {
      const result = applyHunk(lines, hunk)
      if (!result.success) {
        return result
      }
    }

    return {
      success: true,
      newContent: lines.join('\n')
    }
  } catch (error: any) {
    return {
      success: false,
      error: `Failed to apply patch: ${error.message}`
    }
  }
}

/**
 * Apply a single hunk to the lines array
 */
function applyHunk(lines: string[], hunk: DiffHunk): PatchResult {
  const { oldStart, changes } = hunk

  // Convert to 0-based index
  let currentLine = oldStart - 1
  const originalLines = [...lines]

  try {
    for (const change of changes) {
      switch (change.type) {
        case 'context':
          // Verify context line matches
          if (lines[currentLine] !== change.content) {
            return {
              success: false,
              error: `Context mismatch at line ${currentLine + 1}. Expected: "${change.content}", Got: "${lines[currentLine]}"`,
              conflicts: [
                `Line ${currentLine + 1}`,
                `Expected: ${change.content}`,
                `Got: ${lines[currentLine]}`
              ]
            }
          }
          currentLine++
          break

        case 'remove':
          // Verify line to remove matches
          if (lines[currentLine] !== change.content) {
            return {
              success: false,
              error: `Cannot remove line ${currentLine + 1}. Content mismatch.`,
              conflicts: [
                `Line ${currentLine + 1}`,
                `Expected to remove: ${change.content}`,
                `Got: ${lines[currentLine]}`
              ]
            }
          }
          // Remove the line
          lines.splice(currentLine, 1)
          break

        case 'add':
          // Insert new line
          lines.splice(currentLine, 0, change.content)
          currentLine++
          break
      }
    }

    return { success: true }
  } catch (error: any) {
    // Restore original lines on error
    lines.length = 0
    lines.push(...originalLines)

    return {
      success: false,
      error: `Hunk application failed: ${error.message}`
    }
  }
}

/**
 * Apply patch with fuzzy matching (more lenient)
 * Useful when exact line matching fails
 */
export function applyPatchFuzzy(
  originalContent: string,
  patch: string | ParsedDiff,
  fuzziness: number = 2
): PatchResult {
  const diff = typeof patch === 'string' ? parseDiff(patch) : patch
  const lines = originalContent.split('\n')

  for (const hunk of diff.hunks) {
    // Try exact match first
    let result = applyHunk(lines, hunk)

    // If failed, try fuzzy matching
    if (!result.success && fuzziness > 0) {
      result = applyHunkFuzzy(lines, hunk, fuzziness)
    }

    if (!result.success) {
      return result
    }
  }

  return {
    success: true,
    newContent: lines.join('\n')
  }
}

/**
 * Apply hunk with fuzzy line matching
 */
function applyHunkFuzzy(
  lines: string[],
  hunk: DiffHunk,
  fuzziness: number
): PatchResult {
  const { oldStart, changes } = hunk

  // Try to find the hunk at nearby lines
  for (let offset = -fuzziness; offset <= fuzziness; offset++) {
    const testLine = oldStart - 1 + offset
    if (testLine < 0 || testLine >= lines.length) continue

    // Create modified hunk with offset
    const modifiedHunk = {
      ...hunk,
      oldStart: oldStart + offset
    }

    const result = applyHunk([...lines], modifiedHunk)
    if (result.success) {
      // Apply to actual lines
      return applyHunk(lines, modifiedHunk)
    }
  }

  return {
    success: false,
    error: `Could not find matching context for hunk (tried fuzziness: ${fuzziness})`
  }
}

/**
 * Create a reverse patch (for undo)
 */
export function reversePatch(patch: string | ParsedDiff): ParsedDiff {
  const diff = typeof patch === 'string' ? parseDiff(patch) : patch

  return {
    oldFile: diff.newFile,
    newFile: diff.oldFile,
    hunks: diff.hunks.map(hunk => ({
      oldStart: hunk.newStart,
      oldLines: hunk.newLines,
      newStart: hunk.oldStart,
      newLines: hunk.oldLines,
      changes: hunk.changes.map(change => ({
        ...change,
        type: change.type === 'add' ? 'remove' : change.type === 'remove' ? 'add' : 'context'
      }))
    }))
  }
}

/**
 * Preview patch changes without applying
 */
export function previewPatch(
  originalContent: string,
  patch: string | ParsedDiff
): {
  additions: number
  deletions: number
  changes: Array<{ line: number, type: 'add' | 'remove', content: string }>
} {
  const diff = typeof patch === 'string' ? parseDiff(patch) : patch

  let additions = 0
  let deletions = 0
  const changes: Array<{ line: number, type: 'add' | 'remove', content: string }> = []

  for (const hunk of diff.hunks) {
    let line = hunk.newStart

    for (const change of hunk.changes) {
      if (change.type === 'add') {
        additions++
        changes.push({ line, type: 'add', content: change.content })
        line++
      } else if (change.type === 'remove') {
        deletions++
        changes.push({ line, type: 'remove', content: change.content })
      } else {
        line++
      }
    }
  }

  return { additions, deletions, changes }
}
